### PR TITLE
Remove `setTimeout()`. Fix behavior of create menu.

### PR DESCRIPTION
**Pull request by marcaaron**
### Details

@Beamanator @luacmartins I don't think we need this `setTimeout()` anymore. Since this was moved into `WelcomeActions` it seems to open the create menu on iOS just fine.

I removed the `setTimeout()` and tested on a few versions of iOS and it works fine.

### Fixed Issues
$ https://github.com/Expensify/App/issues/5296
Regression caused by: https://github.com/Expensify/App/pull/8692

### Tests (iOS)
1. Sign in as a new user
2. Verify the create menu opens

- [ ] Verify that no errors appear in the JS console

### PR Review Checklist
<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
1. Sign in as a new user
2. Verify the create menu opens

- [ ] Verify that no errors appear in the JS console

#### Web
https://user-images.githubusercontent.com/32969087/166331089-47caebab-2906-47d4-bb24-6bbe00e81fcf.mp4

#### Mobile Web

https://user-images.githubusercontent.com/32969087/166338511-b1f5286e-8fdb-43ce-b1cd-3753098cf140.mp4

#### Desktop

https://user-images.githubusercontent.com/32969087/166338357-af0692ab-8773-4cb7-8169-d4ff409e89f8.mp4

#### iOS

https://user-images.githubusercontent.com/32969087/165857583-28af9f11-cee4-4ac7-b080-4da6c245f296.mp4

#### Android

https://user-images.githubusercontent.com/32969087/166334381-ca27e0b4-3584-41ce-80ab-ef2ddce4fae1.mp4

### CODE REVIEW

**Review by techievivek — DISMISSED** [review]
Works smoothly.

File: src/libs/actions/WelcomeActions.js [hunk 1]
```diff
@@ -9,75 +8,42 @@ import ONYXKEYS from '../../ONYXKEYS';
...
-});
-
 /**
  * Shows a welcome action on first login
  *
  * @param {Object} params
  * @param {Object} params.routes
- * @param {Function} params.hideCreateMenu
+ * @param {Function} params.showCreateMenu
  */
```
Comment (Beamanator) [inline src/libs/actions/WelcomeActions.js:17]: I think we should document the rest of the new params here

```suggestion
 * @param {Function} params.showCreateMenu
 * @param {Boolean} params.isFirstTimeNewExpensifyUser
 * @param {Array} params.allReports
 * @param {Array} params.allPolicies
 */
```

File: src/libs/actions/WelcomeActions.js [hunk 2]
```diff
@@ -9,75 +8,42 @@ import ONYXKEYS from '../../ONYXKEYS';
...
-
-        // It's also possible that we already have a workspace policy. In either case we will not hide the menu but do still want to set the NVP in this case
-        // since the user does not need to create a workspace.
-        if (!Policy.isAdminOfFreePolicy(allPolicies) && !isDisplayingWorkspaceRoute) {
-            hideCreateMenu();
-        }
-    }, 1500);
+function show({
+    routes, showCreateMenu, isFirstTimeNewExpensifyUser, allReports, allPolicies,
+}) {
```
Comment (Beamanator) [inline src/libs/actions/WelcomeActions.js:20]: nab; is a lint rule enforcing doing this on multiple lines? Or personal preference? My personal preference is to keep this 1 line 😅 but def not super worried about it

**Review by Beamanator — DISMISSED** [review]
A few minor comments - tests well!

File: src/libs/actions/WelcomeActions.js (hunk 2 above)
Comment (marcaaron) [reply src/libs/actions/WelcomeActions.js:20]: yeah, it was the linter. I don't have any strong feelings on this though - we can change the linter if it bothers enough people 😄

**Comment by marcaaron** [comment]
Thanks for the comments @Beamanator. Updated!

**Review by kidroca — COMMENTED** [review]
I don't know how others tests this but here's what I did to recreate a related bug

Add a delay for retrieving the `NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER` key
Longer delays recreate the bug more consistently

1. Update the code with the example below and play with the timeout (I managed to reduce the timeout to 1000 and still recreate the bug)
2. Be logged out
3. Log in
4. Observe that even though we're not a first time user "create menu" is opened automatically

#### src/libs/actions/NameValuePair.js
```js
/**
 * Gets the value of an NVP
 *
 * @param {String} name
 * @param {String} onyxKey
 * @param {*} [defaultValue]
 */
function get(name, onyxKey, defaultValue) {
    API.Get({
        returnValueList: 'nameValuePairs',
        name,
    })
        .then((response) => {
            if (name === 'isFirstTimeNewExpensifyUser') {
                setTimeout(() => {
                    console.warn('Overriding isFirstTimeNewExpensifyUser');
                    Onyx.set(onyxKey, false);
                }, 1250);
                return;
            }

            const value = lodashGet(response.nameValuePairs, [name], defaultValue || '');
            Onyx.set(onyxKey, value);
        });
}
```

If the NVP request is slow enough the "Create Menu" will star opened (even if NVP resolves to `false`)(only after login / clean onyx)
Seems like the old 1500 delay was covering this issue as well

We need to either check that we already have a `false` value stored in Onyx, or wait for the NVP request to be over before evaluating whether to show the "create menu" or not

---

The `setTimeout` seem no longer needed otherwise

**Comment by kidroca** [comment]
Maybe we can wait for a boolean value to be added for NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER in Onyx and then open the create menu if necessary? 
1. `componentDidMount` passes show/hide methods to `WellcomeActions` 
2. `WellcomeActions` subscribe to Onyx
3. When boolean value is returned from the subscription it disconnects
4. If the value is `true` the "create menu is displayed"

**Comment by marcaaron** [comment]
>Longer delays recreate the bug more consistently

Thanks @kidroca, just to clarify here the bug is: 

**The create menu is showing when we are not a new user because we must wait to know the true value of `isFirstTimeNewExpensifyUser` from the server**

?

**Comment by marcaaron** [comment]
Does seem like we would probably want to wait for the reports and policies in that case as well 🤔

**Comment by marcaaron** [comment]
Gonna try something else here. I think it will be too confusing to have this logic in the view and we can't get around the fact that we need to wait for data to arrive from the server in this case 😄

**Comment by marcaaron** [comment]
Ok tried a different approach. I'm not sure if it's _super_ great, but I believe it to be more correct than what we had before and here's my full reasoning...

**Fact 1:**

- The "welcome action" **must** only happen until after we have read these values from the server:
   - `isFirstTimeNewExpensifyUser`
   - `allPolicies`
   - `allReports`

I can't think of a way around this right now, but that seems like the primary reason why we have the `setTimeout()`

**Fact 2:**

- We can't trigger the "welcome task" until the Sidebar mounts so calling `WelcomeActions.show()` once in `componentDidMount()` seems correct. This guarantees the modal is ready to open, etc.

**Fact 3:**

- We don't have a great way to detect whether the data we need has loaded from inside the `SidebarScreen`. 

This is a new one for me, but I observed that the value of something like `isReportDataLoading` can be quickly set before the `SidebarScreen` even mounts. If we wanted to move that logic into the view layer it would be kind of ugly and require we do some checks inside `componentDidUpdate()` to see if the data has "loaded" yet. But if it loads before the component even mounts then we'd never be able to know that happened from inside the component because we can:

   - quickly fetch the report data
   - mount the SidebarScreen
   - no way to tell from within SidebarScreen that we have loaded all the reports, policies, or anything else

Sometimes a component needs to know whether we have at least attempted to load data from the server once and that the data it gets from `withOnyx` is the most up to date based on that data load. 

It's not clear how often we might run into this situation, but I feel a more correct approach would be one where the component that needs to know whether this data has loaded would be the one to trigger it. Or we could create some kind of "monitor" that allows any component or lib to know if data has been loaded from the server for the current session.

**Review by luacmartins — DISMISSED** [review]
The new solution works well and LGTM. Should we add videos for all platforms?

**Comment by marcaaron** [comment]
>Should we add videos for all platforms?

I can. It will take me some time to do :D

File: src/libs/actions/Welcome.js [hunk 3]
```diff
@@ -8,17 +8,60 @@ import * as Policy from './Policy';
...
+    },
+});
+
+Onyx.connect({
+    key: ONYXKEYS.IS_LOADING_POLICY_DATA,
+    initWithStoredValues: false,
+    callback: (val) => {
+        isLoadingPolicyData = val;
+        checkOnReady();
+    },
```
Comment (kidroca) [inline src/libs/actions/Welcome.js:58]: Not a big deal, but after `setReady` we no longer need to be connected to these Onyx keys

Comment (kidroca) [reply src/libs/actions/Welcome.js:58]: Actually we might need - for when you logout and login again

File: src/libs/actions/Welcome.js [hunk 4]
```diff
@@ -8,17 +8,60 @@ import * as Policy from './Policy';
 import ONYXKEYS from '../../ONYXKEYS';
 import NameValuePair from './NameValuePair';
 import CONST from '../../CONST';
+import createOnReadyTask from '../createOnReadyTask';
+
+const [isReady, setReady] = createOnReadyTask();
```
Comment (kidroca) [inline src/libs/actions/Welcome.js:13]: Would `isReady` remain resolved if we logout and login with another account?

Comment (kidroca) [reply src/libs/actions/Welcome.js:13]: This is a nice example for a static (libs) vs instance (component) based code

Comment (marcaaron) [reply src/libs/actions/Welcome.js:13]: Good point and nope, but I made this task resettable in another PR so let me see about that...

**Comment by marcaaron** [comment]
Updated, thanks for the spot @kidroca

**Review by luacmartins — APPROVED** [review]

**Comment by marcaaron** [comment]
Gonna be bold and merge this since I think it is blocking the deploy...?

**Comment by marcaaron** [comment]
@sketchydroide my CP staging label didn't seem to work here 🤔  despite [this comment](https://github.com/Expensify/App/issues/8329#issuecomment-1115706724) and [this one](https://github.com/Expensify/App/issues/8777#issuecomment-1116336715) I am pretty sure this needs to staging. But let me know if I missed something.

**Comment by kbecciv** [comment]
@marcaaron
**Login- Create menu is not open**

## Action Performed:
1. Go to staging.new.expensifail.com
2. Sign in as a new user
3. Go back to your email and click on Link
4. Change the URL to staging and set a password for a new user
2. Verify the create menu opens

## Expected Result:
The create menu opens

## Actual Result:
Create menu is not open

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.57.7

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/167160761-44bfa9ba-ab47-4414-83e5-1723305b7ae6.mp4

**Comment by marcaaron** [comment]
@kbecciv doesn't appear to be on staging yet...

**Comment by sketchydroide** [comment]
the code seems to be on staging, let me check if the build for it was deployed

**Comment by kbecciv** [comment]
@sketchydroide We tested on 1.1.57.7 build and it failed

**Pull request merged**